consensus/src/experimental/buffer_manager.rs

```diff
@@ -137,130 +137,104 @@ impl StateManager {
+                _ => unreachable!(),
             }
-            cursor = get_next(&cursor);
         }
-        self.signing_root = cursor;
     }
 
     /// Pop the prefix of buffer items until (including) aggregated_item_cursor
-    /// Prepare an aggregated item and send it to the persisting phase
-    fn try_persisting_blocks(&mut self, aggregated_item_cursor: BufferItemRootType) {
-        let mut cursor = self.buffer.head.as_ref().cloned();
+    /// Send persist request.
+    fn advance_head(&mut self, aggregated_cursor: BufferItemRootType) {
+        let target_block_id = {
+            let item = get_elem(&aggregated_cursor);
+            assert!(item.is_aggregated());
+            item.block_id()
+        };
         let mut blocks_to_persist: Vec<Arc<ExecutedBlock>> = vec![];
 
-        let mut found_item: Option<BufferItem> = None;
-
-        while cursor.is_some() {
-            let buffer_item = take_elem(&cursor);
+        while let Some(item) = self.buffer.pop_front() {
             blocks_to_persist.extend(
-                buffer_item
-                    .get_blocks()
+                item.get_blocks()
                     .iter()
                     .map(|eb| Arc::new(eb.clone()))
                     .collect::<Vec<Arc<ExecutedBlock>>>(),
             );
-            if link_eq(&cursor, &aggregated_item_cursor) {
-                found_item.replace(buffer_item);
-                break;
-            }
-            cursor = get_next(&cursor);
-            self.buffer.pop_front();
-        }
-
-        if let Some(buffer_item) = found_item {
-            if let BufferItem::Aggregated(aggregated) = buffer_item {
-                // send to persisting phase
-                block_on(self.persisting_phase_tx.send(PersistingRequest {
-                    blocks: blocks_to_persist,
-                    commit_ledger_info: aggregated.aggregated_proof,
-                    // we use the last callback
-                    // this is okay because the callback function (from BlockStore::commit)
-                    // takes in the actual blocks and ledger info from the state computer
-                    // the encoded values are references to the block_tree, storage, and a commit root
-                    // the block_tree and storage are the same for all the callbacks in the current epoch
-                    // the commit root is used in logging only.
-                    callback: aggregated.callback,
-                }))
-                .ok();
-            } else {
-                panic!("Bad Aggregated Item Cursor: Item is not aggregated.");
+            if item.block_id() == target_block_id {
+                if let BufferItem::Aggregated(aggregated) = item {
+                    block_on(self.persisting_phase_tx.send(PersistingRequest {
+                        blocks: blocks_to_persist,
+                        commit_ledger_info: aggregated.aggregated_proof,
+                        // we use the last callback
```

Comment:
I'll add a TODO to add some comments at the BlockStore side that the callback needs to be okay if only the last one gets called.